File: CashRegister-May2015/cash_register.go
package main

import (
	"bufio"
	"bytes"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

var denoms = map[int]string{
	10000: "ONE HUNDRED",
	5000:  "FIFTY",
	2000:  "TWENTY",
	1000:  "TEN",
	500:   "FIVE",
	100:   "ONE",
	50:    "HALF DOLLAR",
	25:    "QUARTER",
	10:    "DIME",
	5:     "NICKEL",
	1:     "PENNY",
	0:     "ZERO",
}

type Transaction struct {
	PurchasePrice int
	Cash          int
}

func (t *Transaction) calculateChange() string {
	if t.Cash < t.PurchasePrice {
		return "ERROR"
	}

	if t.Cash == t.PurchasePrice {
		return "ZERO"
	}

	keys := make([]int, 0, len(denoms))
	for k := range denoms {
		keys = append(keys, k)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(keys)))

	diff := t.Cash - t.PurchasePrice

	var buf bytes.Buffer
	for _, i := range keys {
		j := diff / i
		if j > 0 {
			for k := 0; k < j; k++ {
				buf.WriteString(denoms[i])
				diff = diff - i
				if diff > 0 {
					buf.WriteString(",")
				}
			}
			if diff == 0 {
				break
			}
		}
	}
	return buf.String()
}

func main() {
	args := os.Args[1:]

	file, err := os.Open(args[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), ";")
		if len(parts) != 2 {
			log.Printf("Found more than two parts!")
			continue
		}
		pp, err := strconv.ParseFloat(parts[0], 64)
		if err != nil {
			log.Printf("Failed to parse %s", parts[0])
			continue
		}
		c, err := strconv.ParseFloat(parts[1], 64)
		if err != nil {
			log.Printf("Failed to parse %s", parts[0])
			continue
		}
		t := &Transaction{
			PurchasePrice: int(pp * 100),
			Cash:          int(c * 100),
		}
		log.Print(t.calculateChange())
	}
}

File: CashRegister-May2015/sfrisk_javascript/cash_register.js
const cash = {
  'ONE HUNDRED': 100.00,
  'FIFTY': 50.00,
  'TWENTY': 20.00,
  'TEN': 10.00,
  'FIVE': 5.00,
  'TWO': 2.00,
  'ONE': 1.00,
  'HALF DOLLAR': .50,
  'QUARTER': .25,
  'DIME': .10,
  'NICKEL': .05,
  'PENNY': .01
}
var fs  = require("fs");
fs.readFileSync(process.argv[2]).toString().split('\n').forEach(function (line) {
     if (line != "") {
     //do something here
     var money = line.split(";");
      if (money.length == 2) {
          if (money[1] < money[0]) {
              console.log("ERROR");
          } else if (money[1] == money[0]) {
              console.log("ZERO");
          }
          else {
              var difference = (money[1] - money[0]).toFixed(2);
              var change = [];
              for (var key in cash) {
                var t = Math.floor(difference/cash[key]);
                if( t >= 1 ) {
                  difference = (difference - t*cash[key]).toFixed(2);
                  for(i = 0; i < t; i ++) {
                    change.push(key);
                  }
                }
              }
              console.log(change.toString());
          }
      } else {
        console.log("ERROR");
      }
 }
});
